File: core/contracts/src/service.ts
/**
 * The service contract.
 *
 * Every accessibility service implements exactly this interface and declares the
 * resources it needs up front (`requires`). The kernel depends ONLY on this
 * contract — never on a concrete service (Dependency Inversion).
 */

import type { Capability } from './capability.js';
import type { EventBus } from './events.js';
import type { HealthStatus } from './health.js';

export interface ServiceMeta {
  id: string;
  name: string;
  version: string;
}

/** Read-only key/value config + state view injected into a service. */
export interface ConfigStore {
  get<T = unknown>(key: string): T | undefined;
  set<T = unknown>(key: string, value: T): void;
}

/**
 * Everything the kernel injects into a service at load time (Dependency Injection).
 * A service may use the bus and config — but must never reach for another service.
 */
export interface ServiceContext {
  bus: EventBus;
  config: ConfigStore;
  /** The id the kernel registered this service under. */
  selfId: string;
}

/**
 * Lifecycle interface implemented by every service.
 *
 * Lifecycle order: onLoad -> onEnable -> (running) -> onDisable -> onUnload.
 * - onLoad: register, no side effects.
 * - onEnable: acquire resources, start.
 * - onDisable: release resources, stop. Camera/mic leases MUST be released here.
 * - onUnload: final teardown.
 */
export interface AccessibilityService {
  readonly meta: ServiceMeta;

  /** Declarative manifest — drives conflict detection in the Resource Arbiter. */
  readonly requires: Capability[];

  onLoad(ctx: ServiceContext): Promise<void>;
  onEnable(): Promise<void>;
  onDisable(): Promise<void>;
  onUnload(): Promise<void>;
  healthCheck(): HealthStatus;
}

File: core/contracts/python/aah_contracts/service.py
"""The service contract (Python mirror of service.ts).

Python-side services implement :class:`AccessibilityService`. The lifecycle and
``requires`` manifest are identical in shape to the TS contract so the kernel can
treat every service uniformly across the IPC seam.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any

from .capability import Capability
from .health import HealthStatus


@dataclass(frozen=True, slots=True)
class ServiceMeta:
    id: str
    name: str
    version: str


@dataclass(slots=True)
class ServiceContext:
    """Dependencies injected by the kernel host process over IPC.

    ``bus`` is a thin client to the kernel Event Bus; ``config`` is a key/value view.
    A service must never reach for another service — only the bus and config.
    """

    self_id: str
    bus: Any  # IPC bus client; typed concretely by the Python host runtime.
    config: dict[str, Any]


class AccessibilityService(ABC):
    """Lifecycle interface implemented by every Python service.

    Order: on_load -> on_enable -> (running) -> on_disable -> on_unload.
    Camera/mic leases MUST be released in ``on_disable``.
    """

    meta: ServiceMeta
    requires: list[Capability]

    @abstractmethod
    async def on_load(self, ctx: ServiceContext) -> None: ...

    @abstractmethod
    async def on_enable(self) -> None: ...

    @abstractmethod
    async def on_disable(self) -> None: ...

    @abstractmethod
    async def on_unload(self) -> None: ...

    @abstractmethod
    def health_check(self) -> HealthStatus: ...
